### PR TITLE
Add plugin for OpenID connect flow

### PR DESCRIPTION
**Pull request by korycins**
FYI - The name of the branch is incorrect. The changes are generic for openID flow, not exactly for `Auth0`
**I recommend collapsing all yaml files, the review should be easier**

The PR introduces:
- New BasePlugin methods - `authenticate_user`, `external_refresh`, and `external_authenticate`
- New Plugin OpenIDConnect, responsible for handling the external authentication by communicating with external OAuth provider.

Flow for `external_authenticate`:
image

Flow for `external_refresh`:
image

Flow for `authenticate_user`:
image

This PR introduces the part of logic presented in charts as `OpenIDConnectPlugin`

### CODE REVIEW

**Comment by patrys** [comment]
@korycins Where are your famous sequence diagrams? 😉

**Comment by korycins** [comment]
Yes, You are right, let me update and upload them ;)

**Comment by patrys** [comment]
Task linked: [SALEOR-1098 OpenIDConnectPlugin](https://app.clickup.com/t/2549495/SALEOR-1098)

File: saleor/plugins/openid_connect/utils.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+def validate_storefront_redirect_url(storefront_redirect_url: Optional[str]):
+    if not storefront_redirect_url:
+        raise ValidationError(
+            {
+                "redirectUrl": ValidationError(
+                    "Missing redirect url.", code=PluginErrorCode.NOT_FOUND.value
+                )
+            }
+        )
+    validate_storefront_url(storefront_redirect_url)
```
Comment (fowczarek) [inline saleor/plugins/openid_connect/utils.py:44]: This line raise `ValidationError` without `field` and `code`.

File: poetry.lock [hunk 2]
```diff
@@ -155,10 +169,10 @@ description = "The AWS SDK for Python"
...
 optional = false
 python-versions = "*"
-version = "1.15.6"
+version = "1.14.60"
 
 [package.dependencies]
-botocore = ">=1.18.6,<1.19.0"
+botocore = ">=1.17.60,<1.18.0"
 jmespath = ">=0.7.1,<1.0.0"
 s3transfer = ">=0.3.0,<0.4.0"
```
Comment (IKarbowiak) [inline poetry.lock:177]: Why do you downgrade some of the versions of the packages?

Comment (korycins) [reply poetry.lock:177]: I guess, something with merge goes wrong. Will revert it

File: saleor/plugins/openid_connect/utils.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+def get_or_create_user_from_token(claims: CodeIDToken) -> User:
+    user = User.objects.filter(email=claims["email"]).first()
+    if not user:
+        user = User.objects.create(
+            is_active=True,
+            email=claims["email"],
+            first_name=claims.get("given_name", ""),
+            last_name=claims.get("family_name", ""),
+        )
```
Comment (fowczarek) [inline saleor/plugins/openid_connect/utils.py:139]:
```suggestion
    user, _ = User.objects.get_or_create(
        email=claims["email"],
        defaults = {
            "is_active"=True,
            "first_name"=claims.get("given_name", ""),
            'last_name"=claims.get("family_name", ""),
        }
    )
```

**Review by fowczarek — APPROVED** [review]

File: saleor/plugins/openid_connect/plugin.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        configuration = plugin_configuration.configuration
+        configuration = {item["name"]: item["value"] for item in configuration}
+        if plugin_configuration.active:
+            incorrect_fields = []
+            urls_to_validate = {
+                "json_web_key_set_url": configuration["json_web_key_set_url"],
+                "oauth_authorization_url": configuration["oauth_authorization_url"],
+                "oauth_token_url": configuration["oauth_token_url"],
+            }
+            incorrect_fields.extend(get_incorrect_or_missing_urls(urls_to_validate))
```
Comment (IKarbowiak) [inline saleor/plugins/openid_connect/plugin.py:134]: Not needed to define it earlier
```suggestion
            incorrect_fields = get_incorrect_or_missing_urls(urls_to_validate)
```

File: saleor/plugins/openid_connect/utils.py (hunk 3 above)
Comment (korycins) [reply saleor/plugins/openid_connect/utils.py:139]: It is not equal. See the comment from line: 140.  It filters over an email to make sure that the user doesn't exist. If we don't find an email we will create the new user. But if the user already exists we will check if the user is_active

Comment (korycins) [reply saleor/plugins/openid_connect/utils.py:139]: @fowczarek

File: saleor/plugins/openid_connect/plugin.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        return OAuth2Session(
+            client_id=self.config.client_id,
+            client_secret=self.config.client_secret,
+            scope=scope,
+        )
+
+    def external_authentication(
+        self, data: dict, request: WSGIRequest, previous_value
+    ) -> dict:
+        storefront_redirect_url = data.get("redirectUrl")
```
Comment (korycins) [inline saleor/plugins/openid_connect/plugin.py:162]: Check if plugin is avtice here

File: saleor/plugins/openid_connect/utils.py (hunk 3 above)
Comment (patrys) [reply saleor/plugins/openid_connect/utils.py]: @korycins That's not safe, it's possible for another request to create the user right after your first SELECT checks that it does not exist. In that case, `create` will raise an exception. That's what `get_or_create` explicitly guards against.

Comment (korycins) [reply saleor/plugins/openid_connect/utils.py]: @patrys Yeap, thanks. It is already changed

**Review by IKarbowiak — APPROVED** [review]

File: saleor/plugins/openid_connect/plugin.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+            if not configuration["client_id"]:
+                incorrect_fields.append("client_id")
+            if not configuration["client_secret"]:
+                incorrect_fields.append("client_secret")
+            if incorrect_fields:
+                error_msg = (
+                    "To enable a plugin, you need to provide values for the "
+                    "following fields: "
+                )
+                raise ValidationError(
```
Comment (maarcingebala) [inline saleor/plugins/openid_connect/plugin.py:143]: Can we return here a validation error with field name, message, and code for each of these fields? Now, this error is not very useful to the frontend, they can only show that something is wrong but are not able to tell which field failed.

File: saleor/plugins/openid_connect/plugin.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        token = get_token_from_request(request)
+        if not token:
+            return None
+        return get_user_from_access_token(token)
+
+    @convert_error_to_http_response
+    def handle_auth_callback(self, request: WSGIRequest) -> HttpResponse:
+        storefront_redirect_url = request.GET.get("redirectUrl")
+        if not storefront_redirect_url:
+            raise AuthenticationError("Missing get param - redirectUrl")
```
Comment (maarcingebala) [inline saleor/plugins/openid_connect/plugin.py:222]: 👮‍♂️   Doing my grammar police job
```suggestion
            raise AuthenticationError("Missing GET parameter - redirectUrl")
```

Comment (korycins) [reply saleor/plugins/openid_connect/plugin.py:222]: 😞

Comment (korycins) [reply saleor/plugins/openid_connect/plugin.py:222]: @maarcingebala You didn't point the missing dot here 😄

Comment (maarcingebala) [reply saleor/plugins/openid_connect/plugin.py:222]: 😢

**Review by maarcingebala — APPROVED** [review]

**Pull request merged**